core/src/apps/bitcoin/sign_tx/bitcoin.py

```diff
@@ -130,45 +148,94 @@ async def step3_confirm_tx(self) -> None:
-        await helpers.confirm_total(self.total_in - self.change_out, fee, self.coin)
+        if not self.external:
+            await helpers.confirm_total(total, fee, self.coin)
+        else:
+            await helpers.confirm_joint_total(spending, total, self.coin)
 
-    async def step4_serialize_inputs(self) -> None:
+    async def step4_verify_external_inputs(self) -> None:
+        # should come out the same as h_external, checked before continuing
+        h_check = self.create_hash_writer()
+
+        for i in sorted(self.external):
+            progress.advance()
+            txi = await helpers.request_tx_input(self.tx_req, i, self.coin)
+            writers.write_tx_input_check(h_check, txi)
+            prev_amount, script_pubkey = await self.get_prevtx_output(
+                txi.prev_hash, txi.prev_index
+            )
+            if prev_amount != txi.amount:
+                raise wire.DataError("Invalid amount specified")
+
+            if txi.ownership_proof:
+                if not verify_nonownership(
+                    txi.ownership_proof, script_pubkey, b"", self.keychain, self.coin
+                ):
+                    raise wire.DataError("Fake external input")
+            else:
+                verifier = SignatureVerifier(
+                    script_pubkey, txi.script_sig, txi.witness, self.coin
+                )
+
+                verifier.ensure_hash_type(self.get_hash_type(txi))
+
+                tx_digest = await self.get_tx_digest(
+                    i, txi, verifier.public_keys, verifier.threshold, script_pubkey
+                )
+                verifier.verify(tx_digest)
+
+        # check that the inputs were the same as those streamed for confirmation
+        if self.h_external.get_digest() != h_check.get_digest():
+            raise wire.ProcessError("Transaction has changed during signing")
+
+    async def step5_serialize_inputs(self) -> None:
         self.write_tx_header(self.serialized_tx, self.tx, bool(self.segwit))
         write_bitcoin_varint(self.serialized_tx, self.tx.inputs_count)
 
         for i in range(self.tx.inputs_count):
             progress.advance()
-            if i in self.segwit:
+            if i in self.external:
+                await self.serialize_external_input(i)
+            elif i in self.segwit:
                 await self.serialize_segwit_input(i)
             else:
                 await self.sign_nonsegwit_input(i)
 
-    async def step5_serialize_outputs(self) -> None:
+    async def step6_serialize_outputs(self) -> None:
         write_bitcoin_varint(self.serialized_tx, self.tx.outputs_count)
         for i in range(self.tx.outputs_count):
             progress.advance()
             await self.serialize_output(i)
 
-    async def step6_sign_segwit_inputs(self) -> None:
-        any_segwit = bool(self.segwit)
+    async def step7_sign_segwit_inputs(self) -> None:
+        if not self.segwit:
```

Comment:
The progress is finished, so it doesn't break anything. But I can make it future-proof in case a new step is added at some point in the future: cc5d991f655639dff720fc26c26f557db0588449